numpy/core/include/numpy/npy_cpu_dispatch.h

```diff
@@ -0,0 +1,81 @@
+#endif /* NPY_DISABLE_OPTIMIZATION */
+
+/* Multiple declare of prototype with add targeted CPU features as suffix */
+#define NPY_CPU_DISPATCH_DECLARE(FN, ARGS) \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_DECLARE, FN, ARGS) \
+    NPY__CPU_DISPATCH_BASELINE_CALL(__NPY_CPU_DISPATCH_DECLARE_B, FN, ARGS)
+
+/* Same as `NPY_CPU_DISPATCH_DECLARE` but exclude CPU baseline */
+#define NPY_CPU_DISPATCH_DECLARE_XB(FN, ARGS) \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_DECLARE, FN, ARGS)
+
+#define __NPY_CPU_DISPATCH_DECLARE(FNAME, FN, ARGS) NPY_CAT(FN ## _, FNAME) NPY_EXPAND(ARGS);
+/* note: baseline is declared without any suffix */
+#define __NPY_CPU_DISPATCH_DECLARE_B(FN, ARGS) FN NPY_EXPAND(ARGS);
+
+/*
+ * Call a function that suffixed with targeted CPU features after checking the availability in runtime,
+ * the priority depends on the highest interest and it failback to baseline if none of the targeted features are available.
+ */
+#define NPY_CPU_DISPATCH_CALL(FN, ARGS) \
+    if (0) {} \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_CALL, FN, ARGS) \
+    NPY__CPU_DISPATCH_BASELINE_CALL(__NPY_CPU_DISPATCH_CALL_B, FN, ARGS)
+
+/* Same as `NPY_CPU_DISPATCH_CALL` but exclude failback to CPU baseline */
+#define NPY_CPU_DISPATCH_CALL_XB(FN, ARGS) \
+    if (0) {} \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_CALL, FN, ARGS)
+
+#define __NPY_CPU_DISPATCH_CALL(FNAME, FN, ARGS) \
+    else if (NPY_CPU_HAVE(FNAME)) { NPY_CAT(NPY_CAT(FN, _), FNAME) ARGS; }
+#define __NPY_CPU_DISPATCH_CALL_B(FN, ARGS) \
+    else { FN NPY_EXPAND(ARGS); }
+
+/*
+ * Assign a function to a variable that suffixed with targeted CPU features after checking the availability in runtime.
+ * the priority depends on the highest interest and it failback to baseline if none of the targeted features are available.
+*/
+#define NPY_CPU_DISPATCH_ASSIGN(VAR, FN) \
+    if (0) {} \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_ASSIGN, VAR, FN) \
+    NPY__CPU_DISPATCH_BASELINE_CALL(__NPY_CPU_DISPATCH_ASSIGN_B, VAR, FN)
+
+/* Same as `NPY_CPU_DISPATCH_ASSIGN` but exclude failback to CPU baseline */
+#define NPY_CPU_DISPATCH_ASSIGN_XB(VAR, FN) \
+    if (0) {} \
+    NPY__CPU_DISPATCH_CALL(__NPY_CPU_DISPATCH_ASSIGN, VAR, FN)
+
+#define __NPY_CPU_DISPATCH_ASSIGN(FNAME, VAR, FN) \
+    else if (NPY_CPU_HAVE(FNAME)) { VAR = NPY_CAT(NPY_CAT(FN, _), FNAME); }
+#define __NPY_CPU_DISPATCH_ASSIGN_B(VAR, FN) \
+    else { VAR = FN; }
+
+/*
+ * concat `@NAME` with current targeted CPU feature and it should be called from dispatch file,
+ * mainly used to define a function that declared by `NPY_CPU_DISPATCH_DECLARE`
+ * that also be called later by `NPY_CPU_DISPATCH_CALL` or `NPY_CPU_DISPATCH_ASSIGN`
+*/
+#ifdef NPY__CPU_TARGET_MODE
+    #define NPY_CPU_DISPATCH_CURFX(NAME) NPY_CAT(NAME ## _, NPY__CPU_TARGET_CURRENT)
+#else
+    #define NPY_CPU_DISPATCH_CURFX(NAME) NAME
+#endif
+
+#endif /* NPY_CPU_DISPATCH_H_ */
```

Comment:
Is this meant to be all exposed to users who consume numpy's headers, or is it for internal use only when building numpy? We are trying not to expand our public API, so anything that can move to `numpy/core/src/common` should.

Reply:
It will be great to spread all this work for all users, this header is independent not included in any other headers except `simd.h`.
all user has to do is using `ccompiler_opt` to generate their own `_cpu_dispatch.h` and provide the path of it to the used compiler.